### ISSUE TITLE
TM versus GD versus IM versus DD Discussion

### ISSUE THREAD

**Issue by smiths** [JacquesCarette/Drasil#2599]
Discussion about the differences between TMs (Theoretical Models), GDs (General Definitions), IMs (Instance Models) and DDs (Data Definitions) come up all the time when working with Drasil.  I've created this issue so that I can present the current thinking on these definitions and propose my current idea on how to modify the definitions and improve the distinction.

The terminology and examples will all be from the viewpoint of an external reader of the SRS (Software Requirements Specification) documentation.  The distinctions made are from the point of view of the document reader.  I'm leaving for a separate discussion what information should be reflected in the internal Drasil representation of this knowledge and how best to encode this information.

I will first present the definitions I've been using for years.  I no longer think these definitions are adequate, but I won't get into that discussion until after I've presented the mental model that has been used to date.  I'll also define the supporting terminology of Goals and Assumptions.

**Goal**

A goal is a functional objective the system under consideration should achieve. Goals provide criteria for sufficient completeness of a requirements specification and for requirements pertinence. Goals are written abstractly, with a minimal amount of technical language and no equations. They should be understandable by non-domain experts. Goals are refined by the instance models. 

A sample goal from SWHS is "Predict the temperature of the water over time."

**Assumptions**

The assumptions are a refinement of the scope. The scope is general, where the assumptions are specific. All assumptions should be listed, even those that domain experts know so well that they are rarely (if ever) written down.  The assumptions will be used to refine the theoretical models and general definitions.

A sample assumption is "The only form of energy that is relevant for this problem is thermal energy. All other forms of energy, such as mechanical energy, are assumed to be negligible."

**Theoretical Model**

Theoretical models are abstract mathematical equations that will be refined into the instance models that are used to satisfy the goals for the project.  The theoretical models are general.  They will potentially come up in many different physical problems.

[Issue not done - submitting as a way to save it]

**General Definition**

**Instance Model**

**Data Definition**

**Comment by JacquesCarette** [comment]
This is fantastic - definitely spells out lots of terminology. I need some time to have it sink in, and make a proposal.

One note: different disciplines abuse the words 'theory' and 'model' a lot. And these 'disciplines' can be sub-disciplines. The word 'model' means many many different things, and is fraught with danger because of that. Theory is not quite as bad, but also suffers from overloading.

**Comment by smiths** [comment]
Yes, I think I took advantage of the different uses of the words theory and model when I devised the SRS template.  :smile:  It would be nice to nail down the terminology so that we are at least internally consistent with how were use the terms.

**Comment by balacij** [comment]
Thank you for the detailed analysis! It certainly clears things up for me.

Considering what you said regarding the names differing between projects, I wonder if we can merge the models like you mentioned but with a ranked name grouping scheme placed on top of the models so that we can retain the names in areas but change it elsewhere (with explicit explanations of the terminology of course, and a restriction that models of a name group may only reference models of a name group with a rank >= their name group's rank). These name groups would likely need to have some defining factors enforced between them (potentially just whether or not the models are directly used in the resulting programs). I don't know how often this would be useful however.

**Comment by JacquesCarette** [comment]
I'll start with definitions taken from the mathematical side of things.

- **Theory**: Already here there are inconsistencies in the mathematical literature. So let me define two versions:
  - **Theory Presentation**: a set of sorts (or types), function (and sometimes relation) symbols, and a set of axioms. 
  - **Complete Theory**: a theory presentation and all the theorems that can be derived from it.
 
The second is often assumed in logic texts, because it is mathematically elegant, and completely impossible to actually write down. It is what I would call an *ineffective* concept. But it is the correct concept for the 'semantics' or meaning of a theory. Note that it implicitly assumes a background logic with rules of derivations, as a means to establish theorem-hood.

- **Model**: given a particular 'home' (usually set theory), a *model* of a particular *theory* is a way to give a carrier for each sort, actual functions (and relations) for each function symbol, and a proof that each axioms holds.  So a 'theory' could be the 'theory of monoids' (one type, a binary symbol, a unary symbol, and 3 axioms: associativity, left and right unitality), and then 3 models would be (Nat, +, 0), (Nat, *, 1), (String, ++, []).

And then it gets complicated. One can extend a theory with definitions (using the vocabulary of the background logic). These extensions necessarily 'hold' in all models, by rewriting them in terms of the model. Some theories are even 'ground theories': they actually have 0 sorts, 0 functions and 0 axioms, only definitions! If your ambient logic is powerful enough, this happens a lot. What this means is that you already have a large source of models available to you, because the signature of all these definitions is a theory presentation.

A PL analogy is probably good at this point:
- A **Theory Presentation** is a like an ML signature, a Haskell typeclass, a Scala trait or a Java interface.
- A **Model** is like an ML structure, a Haskell instance,  a Scala `extends`, or a Java `implements`
- In most of these languages, it is possible to define (overridable) functions in such a 'Theory'.

When you're working in Haskell (say), you use the same language to define both typeclasses and instances. The instances are full of terms (the stuff on the right-and-side of the `=` for each method of the typeclass that defining an instance requires).

All of this to say that, in some sense, there is more of a spectrum between pure theories and pure models than just these two extremes. On the other hand, there is a very important relation between them: the instance-of relation. A given theory may have many many different models - same with all the PL concepts. That's their very reason to exist!

Now, theories are not unsocial beasts, where you have a small zoo of fixed theories and that's it. [If the PL analogy is to hold, that's rather obvious!]  There are various operations we can do on theories. Some of the go from theories to theories (adding new symbols, new axioms, etc). These kinds of operations are beyond Drasil right now (and were the topic of the PhD theses of Yasmine Sharoda and Musa Al-hassy).

The more important one, for Drasil, is **refinement**: making explicit choices that restrict the number of models. For example, a general theory might have been in n dimensions, but now we fix n to be 3. Models of our refined theory will still be models of the big one, with an additional constraint added. **Refinement** is extremely important for Drasil, as it encapsulates 'design choices' in their purest form. There are many different kinds of refinements, but I won't go into that here.

In theory, a refinement could cause a theory presentation to change quite a lot, by causing many of the embedded expressions to 'simplify'. Very true. And something which will currently be done 'by hand'.

So what does this mean for the topic of this issue?
- **TM** By and large, these correspond to theory presentations.
- **GD** From the description above, these are mostly theory presentations too, but obtained by a refinement step, from either a TM or another GD.
- **IM** seems to be either a model or a theory we don't want to refine anymore. These **IM** have a specific 'form', in that they are constructive: a specific set of symbols are marked as "free" (or "input"), and the rest as "output" [we can deal with parameters later], and the theory is such that all output can be computed from the inputs.
- **DD** these are exactly definitional extensions. The think that is missing from many DD is the context of definition: i.e. what is the background theory in which these make sense. Once we know that, they can be freely added to anything that refined that theory.

Concretely, I suggest the following:
- **TM** be renamed **Theory**. (I would prefer *theory presentation*, but that's overly technical).
- **GD** disappear as a named thing. A **GD** would still exist, but would really be the result of a process akin to `NewThy = ApplyRefinement R1 (ApplyRefinment R2 Thy))`. It could be named 'refined theory'. I don't suggest that exact syntax, it's actually kind of bad. But it's the "right" idea.
- **IM** should probably be called either 'ground theory' or 'terminal theory' or 'computational theory' (or something else). I don't want to use the word 'model' at all, as it's way too overloaded.
- **DD** should be **Definition**. The main thing to change is that it should always be attached to a theory.

The notion of [scientific model](https://en.wikipedia.org/wiki/Scientific_modelling) pointed to by @smiths is in my vocabulary a theory presentation, which is deemed to have a particular relationship with the real world. It is usually somewhat crude, in the sense that there is an unwritten theory that one believes would be 'more accurate' that this is a refinement of, but the larger theory is never written down, as it would be too unwieldly. Sometimes, the assumptions made in this refinement are explicitly written down (ex: no friction).

There are a bunch more technical details, but I think this is enough to get a conversation going.

**Comment by smiths** [comment]
@JacquesCarette this is a great post.  There is plenty to digest here.  Once I've digested it, I'd like to try out your definitions by applying them (manually at first) to the existing case studies.  I find it easier to see how everything fits together through examples.  This will be especially helpful for understanding how a definition is attached to a theory.

**Comment by balacij** [comment]
In the context of #2883, I wonder if an alternative option is to only rewrite the encoding of TMs/GDs/IMs/DDs to use "Theory Presentations" under the hood. Then, we could create 2 variants of our SRS documents, as applicable, where TMs/GDs/DDs are additional definitions which add context to the theories for the SRS documents. I know that @smiths mentioned he had a problem with the "background theories vs refined theories" he prototyped on NoPCM (on the caseStudies repo), so this might be a potential resolution to it, on a case-by-case basis.

**Comment by JacquesCarette** [comment]
As we moved forward with our documents, we realized that the very definition of the various concepts (TM/GD/IM) was a bit too loose. At first, that was fine, but eventually we ran into problems. So I think that a rethink of the concepts is the better way to go.

I do like the idea of having multiple templates, with multiple names for things. However our first template should have 'solid' definitions for the concepts, which is currently not the case. 

I think we're almost at the design stage. I think @smiths is still doing one example by hand to see how it goes, and that will feed into the design.

**Comment by smiths** [comment]
Yes, I am working on an example for the noPCM SRS.

**Comment by JacquesCarette** [comment]
No, 'concrete' does not mean that it can no longer be refined. It only means that it is specific enough that it is possible to give (unique) values to everything under some circumstances.

For example, a theory might have parameters and be considered concrete if we assume that we'll always be operating in an environment where those parameters will have known values. **will have** is key here. Such a theory is still refinable, as we could choose to specialize a parameter **now**.

**Comment by smiths** [comment]
Yes, I agree with @JacquesCarette.  In the SRS for noPCM we can consider the IM as refined to the point where we can generate code.  The code will involve solving an ODE, but we will haven't said what algorithm to use to solve the ODE.  Even after the algorithm is selected we can potentially specialize parameters.

**Comment by balacij** [comment]
Hmm, ok, I think that makes sense. To be clear, does this mean that 'concrete' and 'abstract' are terms that are attached only with respect to a particular target language (for us, largely GOOL)?

Is there anything we can say about the allowed refinement types once a theory has become 'concrete'? For example, do they all ensure that the theory is still 'concrete' and that no new 'target language generation choices' appear?

**Comment by smiths** [comment]
@balacij I consider abstract and concrete as relative concepts.  I think part of the benefit of Drasil is that we don't have to decide "for all time" whether something is abstract or concrete.  We can draw a line at the abstraction level where we want to start and we can say that this is our known starting point.  We don't have to justify it or derive it.  If in the future we find that it is advantageous to be more general, possibly because being more general (abstract) lets us create something that can be used in another context, then we can move back the starting line; we can define a new point of what is known.

The same applies at the concrete end - the currently called "final theories".  The final theories are the point where we say we are done, but we can change that in the future.

I don't think we have any need to identify the absolute "top", the most abstract theory, or an absolute "bottom" of the most concrete theory.  We know when one theory is more abstract or concrete than another, but where we start and finish for any particular problem are determined by practical considerations.

**Comment by JacquesCarette** [comment]
Here is an example: is `a*x^2+b*x+c` a concrete polynomial or not? Well, if we have values for `a,b,c` it is. But more generally so: in a context where we have 'promises' of values for those 3, we can consider it concrete too.

Similarly, if all we know is that "p is a polynomial" then for sure it isn't concrete. It has neither a degree nor coefficients. "p is a degree 2 polynomial" is also abstract - and different from the above!! This is because before we had a concrete handle (i.e. names) for the coefficients, while here we only know they exist (and there are 3 of them).

So indeed 'concrete' and 'abstract' are relative terms, and we don't want to change that.

And so my take: a theory is "concrete enough" if it is possible to generate code from it. You can't generate code from `p`.

**Comment by samm82** [comment]
Since we've been talking about a ["little theories"](https://github.com/smiths/caseStudies/blob/master/CaseStudies/projectile/projectileSRS_RefinedTheories/RefiningTheories/ProjectileLittleTheories.pdf) approach, what should we do with this discussion? Is there anything left here that's worth documenting, or does our current discussion render this "obsolete"?

**Comment by JacquesCarette** [comment]
As it is a wide ranging discussion, it seems quite hard to tell. Especially as some of the knowledge in the above is now understood informally without our group -- but has it been written down? I don't know!

**Comment by smiths** [comment]
This issue probably should have been created in the Discussion forum, rather than as an issue.  However, given that it is here, I don't think we should close it just yet.  Once we start implementing our theory refinements into Drasil, we will likely re-read this issue.

**Comment by samm82** [comment]
I've re-labelled this issue based on these comments (hopefully correctly!)
